### DISCUSSION TITLE
How can I simulate heat diffusion with different materials?

### DISCUSSION THREAD

**Discussion by marcobernardes** [zwicker-group/py-pde#43] [Q&A]
I was wondering if it would be possible to simulate 2D heat diffusion with different materials.
```python
from pde import DiffusionPDE, ScalarField, UnitGrid
grid = UnitGrid([32, 32]) # generate grid
state = ScalarField(grid) # generate initial condition
# set boundary conditions `bc` for all axes
bc_x_left = {"value": 0}
bc_x_right = {"value": 100}
bc_x = [bc_x_left, bc_x_right]
#bc_y_top = {"value": 0}
#bc_y_bottom = {"value": 0}
bc_y_top = {"derivative": 0}
bc_y_bottom = {"derivative": 0}
bc_y = [bc_y_bottom, bc_y_top]
eq = DiffusionPDE(10,bc=[bc_x, bc_y])
result = eq.solve(state, t_range=200, dt=0.005)
result.plot()
```

I think the best way to do this would be to set up the diffusivity values independently. How can I do this?

Many thanks!

**Comment by david-zwicker** [comment]
There are at least two alternatives: Either you [write your own class to implement the PDE](https://py-pde.readthedocs.io/en/latest/manual/advanced_usage.html#custom-pde-classes) or you use the [predefined class `PDE`](https://py-pde.readthedocs.io/en/latest/packages/pde.pdes.html#pde.pdes.pde.PDE) to specify a diffusion equation with variable coefficients. In the latter case, you can [use spatially dependent coefficients](https://github.com/zwicker-group/py-pde/discussions/41) to describe your material properties.

**Reply by marcobernardes** [reply]
Thank you David!

I could implement the Heat Diffusion equation but I don't know how to set customized diffusivity.
How can I do that, please?

Many thanks,

look:

from pde import PDE, DiffusionPDE, ScalarField, UnitGrid, CartesianGrid, MemoryStorage, plot_kymograph

grid = CartesianGrid([[0, 1], [0, 1]], [32, 32]) # generate grid
state = ScalarField(grid, data=0)

# set boundary conditions `bc` for all axes
bc_x_left = {"value": 0}
bc_x_right = {"value": 100}
bc_x = [bc_x_left, bc_x_right]
bc_y_top = {"value": 0}
bc_y_bottom = {"value": 0}
bc_y = [bc_y_bottom, bc_y_top]

# Expanded definition of the PDE
diffusivity = "0.01"

field = ScalarField(grid, data=0.1)
field.add_interpolated([0.5, 0.5], 0.5)
#field.plot()

eq = PDE({"c": f"{diffusivity} * laplace(c)"}, bc=[bc_x, bc_y])

result = eq.solve(state, t_range=10, dt=0.001)
result.plot()

**Reply by david-zwicker** [reply]
Could you please specify the exact equation that you're trying to solve?

**Reply by marcobernardes** [reply]
2d heat diffusion

**Reply by david-zwicker** [reply]
This is not what I meant – please give the exact mathematical formulation with boundary conditions. So far, I do not understand where your problem lies.

**Comment by marcobernardes** [comment]
for instance:


                         T = 1
       
       %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
       %                                        %
       %                                        %
       %    %%%%%%%%%                           %
       %    %       %                           %
       %    %  k1   %           k3              %
       %    %       %                           %
       %    %%%%%%%%%                           %
       %                                        % T = 0
       %                                        %
       %                                        %
       %                        %%%%%%%%%       %
       %                        %       %       %
       %                        %  k2   %       %
       %                        %       %       %
       %                        %%%%%%%%%       %
       %                                        %
       %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

                        T = 0
